ftfy/formatting.py

```diff
@@ -0,0 +1,141 @@
+    cell, because we assume that the monospaced display you're using is
+    designed mostly to display Western characters with CJK characters as the
+    exception. This assumption will go wrong, for example, if your display is
+    actually on a Japanese flip-phone... but let's assume it isn't.
+
+    If this character is a control code, the idea of it having a "width"
+    probably doesn't even apply, because it will probably have some other
+    effect on your terminal. For example, there's no sensible width for a
+    line break. We return the default of 1 in the absence of anything sensible
+    to do there.
+
+    >>> character_width('車')
+    2
+    >>> character_width('A')
+    1
+    >>> character_width('\N{SOFT HYPHEN}')
+    0
+    """
+    if combining(char) != 0:
+        # Combining characters don't advance the cursor; they modify the
+        # previous character instead.
+        return 0
+    elif east_asian_width(char) in 'FW':
+        # Characters designated Wide (W) or Fullwidth (F) will take up two
+        # columns in many Unicode-aware terminal emulators.
+        return 2
+    elif category(char) == 'Cf':
+        # Characters in category Cf are un-printable formatting characters
+        # that do not advance the cursor, such as zero-width spaces or
+        # right-to-left marks.
+        return 0
+    else:
+        return 1
+
+
+def monospaced_width(text):
+    """
+    Return the number of character cells that this string is likely to occupy
+    when displayed in a monospaced, modern, Unicode-aware terminal emulator.
+    We refer to this as the "display width" of the string.
+
+    This can be useful for formatting text that may contain non-spacing
+    characters, or CJK characters that take up two character cells.
+
+    >>> monospaced_width('ちゃぶ台返し')
+    12
+    >>> len('ちゃぶ台返し')
+    6
+    """
+    # NFC-normalize the text first, so that we don't need special cases for
+    # Hangul jamo.
+    text = normalize('NFC', text)
+    return sum([character_width(char) for char in text])
+
+
+def display_ljust(text, width, fillchar=' '):
+    """
+    Return `text` left-justified in a Unicode string whose display width,
+    in a monospaced terminal, should be at least `width` character cells.
+    The rest of the string will be padded with `fillchar`, which must be
+    a width-1 character.
+
+    "Left" here means toward the beginning of the string, which may actually
+    appear on the right in an RTL context. This is similar to the use of the
+    word "left" in "left parenthesis".
+
+    >>> lines = ['Table flip', '(╯°□°)╯︵ ┻━┻', 'ちゃぶ台返し']
+    >>> for line in lines:
+    ...     print(display_ljust(line, 20, '▒'))
+    Table flip▒▒▒▒▒▒▒▒▒▒
+    (╯°□°)╯︵ ┻━┻▒▒▒▒▒▒▒
+    ちゃぶ台返し▒▒▒▒▒▒▒▒
+    """
+    if character_width(fillchar) != 1:
+        raise ValueError("The padding character must have display width 1")
+    text_width = monospaced_width(text)
```

Comment:
Some of these temporaries could be removed without sacrificing clarity, I think.